WebCore::XPath::Parser::Token contains uninitialized fields after construction

ddkilzer opened the pull request:
#### ff290911a4c8c24e3863c1b6f38d3da07b3fcf5b
<pre>
WebCore::XPath::Parser::Token contains uninitialized fields after construction
&lt;<a href="https://bugs.webkit.org/show_bug.cgi?id=274872">https://bugs.webkit.org/show_bug.cgi?id=274872</a>&gt;
&lt;<a href="https://rdar.apple.com/128972339">rdar://128972339</a>&gt;

Reviewed by Chris Dumez.

Merge individual token values into a std::variant&lt;&gt; type named
TokenValue to make sure their initialized state and value are tracked
properly.

* Source/WebCore/xml/XPathParser.cpp:
(WebCore::XPath::Parser::Token::TokenValue): Add.
- Declare std::variant type to hold one of four possible values.
(WebCore::XPath::Parser::Token::Token):
- Replace four instance variables with TokenValue variable.
- Delete default constructor since it is unused.
- Condense constructors taking different value arguments into a single
  constructor taking a TokenValue (std::variant) r-value.
(WebCore::XPath::Parser::Token::string): Add.
(WebCore::XPath::Parser::Token::axis): Add.
(WebCore::XPath::Parser::Token::numericOpcode): Add.
(WebCore::XPath::Parser::Token::equalityTestOpcode): Add.
- Add getter methods to read `value` instance variable.
(WebCore::XPath::Parser::lex):
- Update to use getter methods.

Canonical link: <a href="https://commits.webkit.org/279493@main">https://commits.webkit.org/279493@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/50ced5b0e84198a82a0e44cf7d19413fc235b423

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53637 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32998 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/6150 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56917 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4363 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/55940 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/40478 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/4178 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/43455 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2849 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55734 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/31196 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46370 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/24590 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/28042 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3687 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2518 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/49777 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3868 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58512 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28799 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/3892 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/50864 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/29999 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46534 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/50205 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/30930 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7910 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29776 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/xml/XPathParser.cpp:
```diff
@@ -47,16 +47,31 @@ namespace XPath {
+    Token(int type)
+        : type(type)
+    { }
+    Token(int type, const String& string)
```
Maybe we can take a `const TokenValue&` as parameter and require only a single constructor instead of 4?

ddkilzer replied:
Great idea!  I replaced all four constructors with this:

```
    Token(int type, TokenValue&& value)
        : type(type), value(WTFMove(value))
    { }
```

cdumez commented on Source/WebCore/xml/XPathParser.cpp:
```diff
@@ -47,16 +47,31 @@ namespace XPath {
+    { }
+
+    String& string() { return std::get<String>(value); }
+    Step::Axis axis() { return std::get<Step::Axis>(value); }
```
function could be const.

ddkilzer replied:
You know you can highlight multiple lines by click-and-drag so you only have to write a single comment?  :)

cdumez replied:
Ah! Today I learn :)

cdumez commented on Source/WebCore/xml/XPathParser.cpp:
```diff
@@ -47,16 +47,31 @@ namespace XPath {
+
+    String& string() { return std::get<String>(value); }
+    Step::Axis axis() { return std::get<Step::Axis>(value); }
+    NumericOp::Opcode numericOpcode() { return std::get<NumericOp::Opcode>(value); }
```
function could be const.

cdumez commented on Source/WebCore/xml/XPathParser.cpp:
```diff
@@ -47,16 +47,31 @@ namespace XPath {
+    String& string() { return std::get<String>(value); }
+    Step::Axis axis() { return std::get<Step::Axis>(value); }
+    NumericOp::Opcode numericOpcode() { return std::get<NumericOp::Opcode>(value); }
+    EqTestOp::Opcode equalityTestOpcode() { return std::get<EqTestOp::Opcode>(value); }
```
function could be const.

Commit:
WebCore::XPath::Parser::Token contains uninitialized fields after construction
<https://bugs.webkit.org/show_bug.cgi?id=274872>
<rdar://128972339>

Reviewed by Chris Dumez.

Merge individual token values into a std::variant<> type named
TokenValue to make sure their initialized state and value are tracked
properly.

* Source/WebCore/xml/XPathParser.cpp:
(WebCore::XPath::Parser::Token::TokenValue): Add.
- Declare std::variant type to hold one of four possible values.
(WebCore::XPath::Parser::Token::Token):
- Replace four instance variables with TokenValue variable.
- Delete default constructor since it is unused.
- Condense constructors taking different value arguments into a single
  constructor taking a TokenValue (std::variant) r-value.
(WebCore::XPath::Parser::Token::string): Add.
(WebCore::XPath::Parser::Token::axis): Add.
(WebCore::XPath::Parser::Token::numericOpcode): Add.
(WebCore::XPath::Parser::Token::equalityTestOpcode): Add.
- Add getter methods to read `value` instance variable.
(WebCore::XPath::Parser::lex):
- Update to use getter methods.

Canonical link: https://commits.webkit.org/279493@main
